File: 2023/03A/main.zig
const std = @import("std");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    var stdin = std.io.bufferedReader(std.io.getStdIn().reader());
    var stdout = std.io.bufferedWriter(std.io.getStdOut().writer());

    var grid = std.ArrayList(std.ArrayList(u8)).init(allocator);
    defer grid.deinit();
    defer for (grid.items) |line| {
        line.deinit();
    };

    var lines_done = false;
    while (!lines_done) {
        try grid.append(std.ArrayList(u8).init(allocator));
        stdin.reader().streamUntilDelimiter(grid.items[grid.items.len - 1].writer(), '\n', null) catch |err| switch (err) {
            error.EndOfStream => lines_done = true,
            else => return err,
        };
    }

    var total: u32 = 0;

    for (0.., grid.items) |row, line| {
        const row_lwr = @max(1, row) - 1;
        const row_upr = @min(row + 2, grid.items.len);
        var col: usize = 0;
        while (col < line.items.len) {
            const lwr = col;
            while (col < line.items.len and std.ascii.isDigit(line.items[col])) {
                col += 1;
            }
            const upr = col;
            if (lwr == upr) {
                col += 1;
                continue;
            }
            const val = try std.fmt.parseInt(u32, line.items[lwr..upr], 10);
            const col_lwr = @max(1, lwr) - 1;
            const col_upr = @min(upr + 1, line.items.len);
            var search_done = false;
            for (row_lwr..row_upr) |r| {
                for (col_lwr..col_upr) |c| {
                    if (grid.items[r].items[c] != '.' and !std.ascii.isDigit(grid.items[r].items[c])) {
                        search_done = true;
                        total += val;
                    }
                    if (search_done) break;
                }
                if (search_done) break;
            }
        }
    }

    try stdout.writer().print("{d}\n", .{total});
    try stdout.flush();
}

File: 2023/03A/main.rs
use std::cmp::{max, min};
use std::io;

fn main() {
    let stdin = io::stdin();
    let grid = stdin
        .lines()
        .map(|l| l.unwrap().chars().collect::<Vec<char>>())
        .collect::<Vec<_>>();
    let mut total: u32 = 0;
    for row in 0..grid.len() {
        let row_lwr = max(1, row) - 1;
        let row_upr = min(row + 2, grid.len());
        for col in 0..grid[row].len() {
            if col > 0 && grid[row][col - 1].is_digit(10) {
                continue;
            }
            let len = grid[row][col..]
                .iter()
                .take_while(|c| c.is_digit(10))
                .count();
            if len == 0 {
                continue;
            }
            let val = grid[row][col..col + len]
                .iter()
                .collect::<String>()
                .parse::<u32>()
                .unwrap();
            let col_lwr = max(1, col) - 1;
            let col_upr = min(col + len + 1, grid[row].len());
            'search: for r in row_lwr..row_upr {
                for c in col_lwr..col_upr {
                    if grid[r][c] != '.' && grid[r][c].is_ascii_punctuation() {
                        total += val;
                        break 'search;
                    }
                }
            }
        }
    }
    println!("{}", total);
}
